spec.md

```diff
@@ -2091,6 +2091,17 @@ message NodeStageVolumeRequest {
   // This field is OPTIONAL and MUST match the volume_context of the
   // volume identified by `volume_id`.
   map<string, string> volume_context = 6;
+
+  // If SP has VOLUME_MOUNT_GROUP node capability and CO provides
+  // this field then SP MUST ensure that volume is mounted with
+  // provided volume_mount_group and all files and directories
```

Comment:
I'm concerned that the wording "all files and directories within the volume are readable and writable by ..." leaks implementation detail, thinking that it should be removed. In terms of guarantees, the spec should aim no higher than what's provided by `mount` (and the OS driver implementation). Driver implementations of `mount` are already a black box and it feels odd to layer additional requirements on top of a underspecified interface. Furthermore, just because k8s wants to implement a recursive `chown` doesn't mean that it's appropriate for all CSI driver implementations, or all CO implementations/deployments.

Also, removing the recursive `chown` wording makes this more interesting for block volumes. Perhaps ownership means something for those volumes too? And if so, do we really want to tie this to "mount", or maybe we need a different word if "mount" is only for fs-type volumes (and vs block-type volumes).

Reply:
> Furthermore, just because k8s wants to implement a recursive chown doesn't mean that it's appropriate for all CSI driver implementations, or all CO implementations/deployments.

Just to clarify - if a driver has this capability, k8s WILL not `chown/chmod` files belonging to this volume. In stead the plan is - k8s will assume SP has done its job and volume has right permissions after nodestage/nodepublish. Also - the reason we need this as a mount flag is mainly because in many cases k8s/CO *can not* recursively `chown` files (for example - cifs mount without unix extensions).

> 'm concerned that the wording "all files and directories within the volume are readable and writable by ..." leaks implementation detail, thinking that it should be removed. 
Yes the wording does gets into implementation details. My reasoning as originally discussed with @bswartz was - we *do not* want SPs to use this field to recursively chown/chmod files at all. Because chown/chmod is very much CO specific thing and we actively wanted to prevent SPs from doing that.  When using `chown` for example - newly written files on the volume may not be group writable and hence would violate the CSI spec.  But I can see - how this leaks into mount implementation details too.

Reply:
Still not happy with this wording:
> ... then SP MUST ensure that volume is mounted with provided volume_mount_group and all files and directories within the volume are readable and writable by the provided volume_mount_group.

(a) what does "mounted with provided volume_mount_group" actually mean?
(b) all files + dirs readable/writable sounds like a recursive chown to me, unless this is a validation check? which doesn't really make sense in this context.

I'd like to take a stab at some alt wording, which attempts to decrease the scope of what CSI is willing to promise:

"... then the SP MUST ensure that the `volume_mount_group` parameter is passed as the group (primary group, if possible) identifier to the underlying OS mount system call, with the understanding that the set of available mount call parameters and/or mount implementations may vary across operating systems. Additionally, new file and/or directory entries written to the underlying filesystem SHOULD be permission-labeled in such a manner, unless otherwise modified by a workload, that they are both readable and writable by said mount group identifier."

.. which tries to get away from saying that everything on a preexisting volume will be read/write-able (which is gets into recursive chown territory), and also tries to cut to the chase of the requirement that I think i'm hearing - that newly written blobs are the concern here, and they should (by default) be labeled with the group ID presented here.

also, if a workload chown's files on the volume such that it can no longer read them .. well, tough luck. CSI is not a data plane API. it's up to the SP implementation and/or OS to restrict such operations if that behavior is undesireable. CSI should not promise anything w/ respect to what I/O operations a workload is allowed to execute (because CSI is a control plane API).

Reply:
I think updated wording looks fine. The only question I have is - why primary group if possible? k8s at least will never pass primary group of the workload.

Reply:
I was thinking some more about it and I agree that your updated wording meets the requirement and everything, but one more thing which has been bothering me somewhat is - we don't really need to call out permissions of new files/directories. The wording does imply that - chown/chmod may not be enough but what we really want SPs to use is - mount options and hoping that mount operation with provided group identifier is enough to apply necessary permissions. 

So I am thinking two alternative wordings:

"...then the SP MUST ensure that the volume_mount_group parameter is passed as the group identifier to the underlying OS mount system call,
with the understanding that this would result in volume being readable(if published as readonly) or  readable and writable by said mount group identifier. The set of available
mount call parameters and/or mount implementations may vary across operating systems." --> This covers **readonly** interaction as well.

Or something like:

"... then the SP MUST ensure that the volume_mount_group parameter is passed as the group identifier to the underlying OS mount system call, with the understanding that the set of available mount call parameters and/or mount implementations may vary across operating systems. Additionally, it is expected file and/or directory entries on the underlying filesystem SHOULD be permission-labeled in such a manner, unless otherwise modified by a workload, that they are either readable(in case of readonly volumes) or both readable and writable by said mount group identifier."

Reply:
My wording says the permission labeling (ala the group option) targets *new* file(s)/dir(s); your wording doesn't include the "new", which in my mind translates to "recursive chown".

If you did not intend to drop "new" from the wording, please add it back. Because maybe you're only trying to amend my wording to account for (a) readable files in the read-only case, or else; (b) read/write-able files otherwise?

Reply:
My main concern with accounting for "new" files is - are we being too specific? A SP may violate this and still satisfy what CO expects. But to be clear we do not have that problem at least with Azurefile/CIFS-Samba mounts. It works exactly like how you worded and it should be fine. So may be I am over-thinking this.

Reply:
alright - I have incorporated this wording.  I thought again about readonly case and I do not think it is important to call that out here again, because readonly can be enforced in different ways than by combining with gid mount options.